### PR TITLE
types: use switch table to save memory

### PR DESCRIPTION
**Pull request by liufuyang**
### What problem does this PR solve? 
To deal with this one:
https://github.com/pingcap/tidb/issues/9642

### What is changed and how it works?
Instead of defining a map, some witch tables in functions are used, and Go compiler will inline it. With this way, there is no memory allocation at all.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by liufuyang** [comment]
@ngaut Can you take a look is this what you want to do or I miss understood the issue. 
I am not so sure about the function name is good enough or not.
Also, I am not sure how to deal with the `default` case in switch. Can you give me some ideas? (I simply return 0 on default case for now just to make the code can compile and run...)

**Comment by ngaut** [comment]
Good job.

File: types/datum.go [hunk 1]
```diff
@@ -1409,8 +1409,8 @@ func (d *Datum) ToInt64(sc *stmtctx.StatementContext) (int64, error) {
 }
 
 func (d *Datum) toSignedInteger(sc *stmtctx.StatementContext, tp byte) (int64, error) {
-	lowerBound := SignedLowerBound[tp]
-	upperBound := SignedUpperBound[tp]
+	lowerBound := ByteToSignedLowerBound(tp)
```
Comment (ngaut) [inline types/datum.go]: Could you remove the origin definition of SignedLowerBound?

Comment (liufuyang) [reply types/datum.go]: @ngaut Do you mean delete `SignedLowerBound` from `types/convert.go`? I remember I tried to search it around and it seems that `SignedLowerBound` used in multiple places. Do you think I should replace them all or?

File: types/convert.go [hunk 2]
```diff
@@ -67,6 +67,42 @@ var SignedLowerBound = map[byte]int64{
...
+	case mysql.TypeShort:
+		return math.MinInt16
+	case mysql.TypeInt24:
+		return mysql.MinInt24
+	case mysql.TypeLong:
+		return math.MinInt32
+	case mysql.TypeLonglong:
+		return math.MinInt64
+	default:
+		panic("Input byte is not a mysql type")
```
Comment (coocood) [inline types/convert.go]: ditto

File: types/datum.go (hunk 1 above)
Comment (coocood) [reply types/datum.go]: @liufuyang 
I think we should replace them all, so we can completely resolve this issue.

File: types/convert.go [hunk 3]
```diff
@@ -67,6 +67,42 @@ var SignedLowerBound = map[byte]int64{
...
+	case mysql.TypeShort:
+		return math.MaxInt16
+	case mysql.TypeInt24:
+		return mysql.MaxInt24
+	case mysql.TypeLong:
+		return math.MaxInt32
+	case mysql.TypeLonglong:
+		return math.MaxInt64
+	default:
+		panic("Input byte is not a mysql type")
```
[the remark this answers is not in the document]
Comment (ngaut) [reply types/convert.go]: If the caller pass the wrong argument, there would be a silence error, i think panic is a better choice, it should never run into the default branch.

**Comment by liufuyang** [comment]
@ngaut @coocood I have replaced/removed those two, plus an extra one. Please take a look :)

File: types/convert.go [hunk 4]
```diff
@@ -37,34 +37,64 @@ func truncateStr(str string, flen int) string {
...
-	mysql.TypeTiny:     math.MaxUint8,
-	mysql.TypeShort:    math.MaxUint16,
-	mysql.TypeInt24:    mysql.MaxUint24,
-	mysql.TypeLong:     math.MaxUint32,
-	mysql.TypeLonglong: math.MaxUint64,
-	mysql.TypeBit:      math.MaxUint64,
-	mysql.TypeEnum:     math.MaxUint64,
-	mysql.TypeSet:      math.MaxUint64,
+// ByteToUnsignedUpperBound indicates the max uint64 values of different mysql types.
+func ByteToUnsignedUpperBound(b byte) uint64 {
```
Comment (coocood) [inline types/convert.go]: The parameter is not general byte, its integer type.
I think the name should be `IntergerUnsignedUpperBound(intType byte)`.

File: types/convert.go [hunk 5]
```diff
@@ -37,34 +37,64 @@ func truncateStr(str string, flen int) string {
...
 
-// SignedLowerBound indicates the min int64 values of different mysql types.
-var SignedLowerBound = map[byte]int64{
-	mysql.TypeTiny:     math.MinInt8,
-	mysql.TypeShort:    math.MinInt16,
-	mysql.TypeInt24:    mysql.MinInt24,
-	mysql.TypeLong:     math.MinInt32,
-	mysql.TypeLonglong: math.MinInt64,
+// ByteToSignedLowerBound indicates the min int64 values of different mysql types.
+func ByteToSignedLowerBound(b byte) int64 {
```
Comment (coocood) [inline types/convert.go]: `IntergerSignedLowerBound(intType byte)`

File: types/convert.go [hunk 6]
```diff
@@ -37,34 +37,64 @@ func truncateStr(str string, flen int) string {
...
 
-// SignedUpperBound indicates the max int64 values of different mysql types.
-var SignedUpperBound = map[byte]int64{
-	mysql.TypeTiny:     math.MaxInt8,
-	mysql.TypeShort:    math.MaxInt16,
-	mysql.TypeInt24:    mysql.MaxInt24,
-	mysql.TypeLong:     math.MaxInt32,
-	mysql.TypeLonglong: math.MaxInt64,
+// ByteToSignedUpperBound indicates the max int64 values of different mysql types.
+func ByteToSignedUpperBound(b byte) int64 {
```
Comment (coocood) [inline types/convert.go]: `IntegerSignedUpperBound(intType byte)`

File: types/convert.go (hunk 4 above)
Comment (liufuyang) [reply types/convert.go]: @coocood May I ask why you say it is an `integer type`?

I basically replaced `var UnsignedUpperBound = map[byte]uint64{` to this function. And `UnsignedUpperBound` is a map with key as `byte`, shouldn't I use `byte` as type here as well? 🤔

Comment (liufuyang) [reply types/convert.go]: Oh, I see you are talking bout the name not the type... Aha. Sorry I miss read your input. I will use `intType` as the name then :)

**Comment by liufuyang** [comment]
@coocood Code fixed as you suggested :)

**Comment by ngaut** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Comment by liufuyang** [comment]
Interesting, I created a simple benchmark function:
```
func BenchmarkDatumToInt64(b *testing.B) {
	d := NewDatum("0")
	sc := new(stmtctx.StatementContext)
	sc.IgnoreTruncate = true

	for n := 0; n < b.N; n++ {
		d.ToInt64(sc)
	}
}
```
Result is:
```
Before:
BenchmarkDatumToInt64-8   	10000000	       162 ns/op

After:
BenchmarkDatumToInt64-8   	20000000	       110 ns/op
```

**Pull request merged**